packages/core/utils/lib/async.js

```diff
@@ -0,0 +1,55 @@
+'use strict';
+
+const { chunk } = require('lodash/fp');
+
+function pipeAsync(...methods) {
+  return async (data) => {
+    let res = data;
+
+    for (const method of methods) {
+      res = await method(res);
+    }
+
+    return res;
+  };
+}
+
+/**
+ * @type { import('./async').mapAsync }
+ */
+function mapAsync(promiseArray, { concurrency = Infinity } = {}) {
+  const appliedConcurrency = concurrency > promiseArray.length ? promiseArray.length : concurrency;
+  const promiseArrayChunks = chunk(appliedConcurrency)(promiseArray);
+
+  return async (iteratee) => {
+    return promiseArrayChunks.reduce(async (prevChunksPromise, chunk, chunkIndex) => {
+      // Need to await previous promise in order to respect the concurrency option
+      const prevChunks = await prevChunksPromise;
+      // As chunks can contain promises, we need to await the chunk
+      const awaitedChunk = await Promise.all(chunk);
+      const transformedPromiseChunk = await Promise.all(
+        // Calculating the index based on the original array, we do not want to have the index of the element inside the chunk
+        awaitedChunk.map((value, index) => iteratee(value, chunkIndex * appliedConcurrency + index))
+      );
+
+      return prevChunks.concat(transformedPromiseChunk);
+    }, Promise.resolve([]));
+  };
+}
+
+/**
+ * @type { import('./async').reduceAsync }
+ */
+function reduceAsync(promiseArray) {
```

Comment:
I don't understand why it's a `promiseArray` and not just an `array`? The only change compared to the original reduce is that the `iteratee` can be async but that it, it doesn't impact the array on which the reduce is applied.

Reply:
Maybe the wording can be reviewed, but `mixedArray` wasn't feeling right either.
What I wanted to say in this variable is that you can have both promises and values inside the array passed to this function.

Reply:
Why can there be promises into it?

Reply:
Because developers can have an array of anything, including Promises. As this function is to be used in an async environment, it won't surprise me if someone wants to provided some promises inside the inputs array.

Reply:
I don't see the usecase for an array of launched Promises, I've never seen that. If it's a mistake from the dev to have an array like that, I don't think we should fix his/her mistake by supporting this possibility, that would enforce a bad practice.